tycho/db/event/__init__.py

```diff
@@ -43,9 +49,18 @@ def __init__(self, collection):
 
     async def update_by_id(self, id, update_doc: ModelEvent, insert: bool = False):
         new_data = serialize_to_db_event(update_doc)
-        result = await self.collection.replace_one(
-            {"_id": id}, new_data, upsert=insert)
-        return result
+
+        retries = 1
+        while retries >= 0:
```

Comment:
this can result in an infinite loop if the exception raised is not DuplicateKeyError, since the retries don't decrement there.

How about a global catch-all that decrements retries instead?

Reply:
The occurrence of any exception other than DuplicateKeyError will directly be raised, thereby breaking the loop and returning the control to the callee. So, it might not go into an infinite loop. Do let me know if I have got it right.

Reply:
ah, good point.

Reply:
Ok, that makes sense. There's probably a more robust strategy here (like CAS) https://stackoverflow.com/questions/11991167/in-mongo-any-way-to-do-check-and-setting-like-atomic-operation

I think you can do this by running replace_one with the full old data, running replace_one, then retrying if it doesn't match. That would in general be more robust, but I don't see why this wouldn't work.

Reply:
Used `find_one_and_replace` operation instead of `replace_one`

Reply:
I don't think this helps, as the issue notes that findAndModify is also affected, in the case that upsert is True:

https://jira.mongodb.org/browse/SERVER-14322

But again, I think this retry works well in most cases, and the variable timeout helps with worst case robustness.